flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/ArrayAggFunction.java

```diff
@@ -0,0 +1,194 @@
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Runtime
+    // --------------------------------------------------------------------------------------------
+
+    /** Accumulator for ARRAY_AGG with retraction. */
+    public static class ArrayAggAccumulator<T> {
+        public LinkedList<T> list;
+        public LinkedList<T> retractList;
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            ArrayAggAccumulator<?> that = (ArrayAggAccumulator<?>) o;
+            return Objects.equals(list, that.list) && Objects.equals(retractList, that.retractList);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(list, retractList);
+        }
+    }
+
+    @Override
+    public ArrayAggAccumulator<T> createAccumulator() {
+        final ArrayAggAccumulator<T> acc = new ArrayAggAccumulator<>();
+        acc.list = new LinkedList<>();
+        acc.retractList = new LinkedList<>();
+        return acc;
+    }
+
+    public void accumulate(ArrayAggAccumulator<T> acc, T value) throws Exception {
+        if (value == null) {
+            if (!ignoreNulls) {
+                acc.list.add(null);
+            }
+        } else {
+            acc.list.add(value);
+        }
+    }
+
+    public void retract(ArrayAggAccumulator<T> acc, T value) throws Exception {
+        if (value != null) {
+            if (!acc.list.remove(value)) {
+                acc.retractList.add(value);
+            }
+        }
+    }
+
+    public void merge(ArrayAggAccumulator<T> acc, Iterable<ArrayAggAccumulator<T>> its)
+            throws Exception {
+        for (ArrayAggAccumulator<T> otherAcc : its) {
+            // merge list of acc and other
+            List<T> buffer = new ArrayList<>();
+            for (T element : acc.list) {
+                buffer.add(element);
+            }
+            for (T element : otherAcc.list) {
+                buffer.add(element);
+            }
+            // merge retract list of acc and other
+            List<T> retractBuffer = new ArrayList<>();
+            for (T element : acc.retractList) {
+                retractBuffer.add(element);
+            }
+            for (T element : otherAcc.retractList) {
+                retractBuffer.add(element);
+            }
+
+            // merge list & retract list
+            List<T> newRetractBuffer = new ArrayList<>();
+            for (T element : retractBuffer) {
+                if (!buffer.remove(element)) {
+                    newRetractBuffer.add(element);
+                }
+            }
+
+            // update to acc
+            acc.list.clear();
+            acc.list.addAll(buffer);
```

Comment:
can't we populate `acc.list` in a single go? The current approach does make sense with a `ListView` but it does not with a `LinkedList` kept in memory.

I believe we don't need the intermediate `buffer` and `newRetractBuffer`